### PR TITLE
perf: Enable CA1815 (override Equals on value types)

### PR DESCRIPTION
**Pull request by Youssef1313**
[CA1815 documentation](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1815)

GitHub Issue (If applicable): Part of #6882

Benchmark:

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
    <OutputType>Exe</OutputType>
 </PropertyGroup>
  <PropertyGroup>
    <PlatformTarget>AnyCPU</PlatformTarget>
    <DebugType>pdbonly</DebugType>
    <DebugSymbols>true</DebugSymbols>
    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
    <Optimize>true</Optimize>
    <Configuration>Release</Configuration>
    <IsPackable>false</IsPackable>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="BenchmarkDotNet" Version="0.12.1" />
    <PackageReference Include="BenchmarkDotNet.Diagnostics.Windows" Version="0.12.1"/>
  </ItemGroup>
</Project>
```

```csharp
using System;
using BenchmarkDotNet;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

namespace testbench
{
    public partial struct FlowLayoutAnchorInfo
    {
        internal FlowLayoutAnchorInfo(in int index, in double offset)
        {
            Index = index;
            Offset = offset;
        }
        public int Index;

        public double Offset;
    }

    public partial struct FlowLayoutAnchorInfo2
    {
        internal FlowLayoutAnchorInfo2(in int index, in double offset)
        {
            Index = index;
            Offset = offset;
        }
        public int Index;

        public double Offset;

        public override bool Equals(object obj)
        {
            return obj is FlowLayoutAnchorInfo2 info &&
                   Index == info.Index &&
                   Offset == info.Offset;
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(Index, Offset);
        }

        public static bool operator ==(FlowLayoutAnchorInfo2 left, FlowLayoutAnchorInfo2 right)
        {
            return left.Equals(right);
        }

        public static bool operator !=(FlowLayoutAnchorInfo2 left, FlowLayoutAnchorInfo2 right)
        {
            return !(left == right);
        }
    }

    public class Program
    {
        public static void Main(string[] args)
        {
            var summary = BenchmarkRunner.Run(typeof(Program).Assembly);
        }
    }


    public class Benchmarks
    {
        [Benchmark]
        public bool Scenario1()
        {
            var a = new FlowLayoutAnchorInfo(0, 0);
            var b = new FlowLayoutAnchorInfo(5, 0);
            return a.Equals(b);
        }

        [Benchmark]
        public bool Scenario2()
        {
            var a = new FlowLayoutAnchorInfo(0, 0);
            var b = new FlowLayoutAnchorInfo(0, 0);
            return a.Equals(b);
        }

        [Benchmark]
        public bool Scenario3()
        {
            var a = new FlowLayoutAnchorInfo2(0, 0);
            var b = new FlowLayoutAnchorInfo2(5, 0);
            return a.Equals(b);
        }

        [Benchmark]
        public bool Scenario4()
        {
            var a = new FlowLayoutAnchorInfo2(0, 0);
            var b = new FlowLayoutAnchorInfo2(0, 0);
            return a.Equals(b);
        }
    }
}
```

image

## PR Type

What kind of change does this PR introduce?

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.Foundation/HResult.cs
Comment (spouliot) [inline src/Uno.Foundation/HResult.cs:12]: I'm not sure if binary compatibility is needed here, since adding those member will have an impact.

If it's fine (to add member) then it would be better to add `IEquatable<HResult>` and have the operation uses the strongly typed `Equals(HResult)` method.

File: src/Uno.Foundation/HResult.cs
Comment (spouliot) [inline src/Uno.Foundation/HResult.cs:20]: same

File: src/Uno.UI/Microsoft/UI/Xaml/Controls/Repeater/FlowLayoutAnchorInfo.cs
Comment (spouliot) [inline src/Uno.UI/Microsoft/UI/Xaml/Controls/Repeater/FlowLayoutAnchorInfo.cs:33]: same

File: src/Uno.UI/Microsoft/UI/Xaml/Controls/Repeater/FlowLayoutAnchorInfo.cs
Comment (spouliot) [inline src/Uno.UI/Microsoft/UI/Xaml/Controls/Repeater/FlowLayoutAnchorInfo.cs:38]: same

File: src/Uno.UWP/ApplicationModel/PackageVersion.cs
Comment (spouliot) [inline src/Uno.UWP/ApplicationModel/PackageVersion.cs:5]: 😄

File: src/Uno.UWP/ApplicationModel/PackageVersion.cs
Comment (spouliot) [inline src/Uno.UWP/ApplicationModel/PackageVersion.cs:30]: You should reverse the calls.

The linker/trimmer won't be able to remove `IEquatable<PackageVersion>.Equals(PackageVersion)` even if unused.

However it would be able to remove the `==` and/or `!=` operators... unless they are used (like in `Equals` here).

IOW it's more efficient to implement the comparison inside `Equals(PackageVersion)` and have
* `Equals(object)` call it;
* `==` call it;
* `!=` call it

File: src/Uno.UWP/Devices/Geolocation/BasicGeoposition.cs
Comment (spouliot) [inline src/Uno.UWP/Devices/Geolocation/BasicGeoposition.cs:13]: same

File: src/Uno.UWP/Gaming/Input/GamepadReading.cs
Comment (spouliot) [inline src/Uno.UWP/Gaming/Input/GamepadReading.cs:70]: same

**Review by spouliot — CHANGES_REQUESTED** [review]

File: src/Uno.Foundation/HResult.cs
Comment (Youssef1313) [reply src/Uno.Foundation/HResult.cs:12]: I didn't know that adding an operator affects binary compatibility.

@jeromelaban Are we okay with that?

File: src/Uno.UWP/ApplicationModel/PackageVersion.cs
Comment (Youssef1313) [reply src/Uno.UWP/ApplicationModel/PackageVersion.cs:5]: I added the interface here because that's what I found when in the decompiled sources from Windows head.

image

You'll find also that I just copied what I found in the decompiled sources regarding how `Equals` methods and `==` are implemented.

Also their `GetHashCode` is very prune to collisions, but I copied it as-is too.
It's probably fine to diverge from WinUI implementation here, what do you think?

File: src/Uno.UWP/ApplicationModel/PackageVersion.cs
Comment (Youssef1313) [reply src/Uno.UWP/ApplicationModel/PackageVersion.cs:30]: As mentioned in the comment above, I just copied whatever I found in the decompiled sources of `Microsoft.Windows.SDK.NET.dll` (10.0.22000.25), but I think we can safely diverge from Windows implementation?

Comment (jeromelaban) [reply src/Uno.UWP/ApplicationModel/PackageVersion.cs:30]: It would make sense to make the change to be linker friendly.

File: src/Uno.UWP/ApplicationModel/PackageVersion.cs
Comment (jeromelaban) [reply src/Uno.UWP/ApplicationModel/PackageVersion.cs:5]: I wonder if `PackageVersion` should be a `record`, as it would create most of it automatically?

File: src/Uno.Foundation/HResult.cs
Comment (jeromelaban) [reply src/Uno.Foundation/HResult.cs:12]: For this particular class this is not a problem, it's not used that much (if at all). Still, I don't see how this is breaking, unless it is because it is a struct? It would be behavior breaking though, because the operator would not be used by older code.

File: src/Uno.UWP/ApplicationModel/PackageVersion.cs
Comment (spouliot) [reply src/Uno.UWP/ApplicationModel/PackageVersion.cs:5]: > It's probably fine to diverge from WinUI implementation here, what do you think?

Yes. Code should not depend on this level of implementation details. It's part of why net6 added `HashCode.Combine` (and friends) to ensure you can't expect the same values between runs.

Comment (Youssef1313) [reply src/Uno.UWP/ApplicationModel/PackageVersion.cs:5]: I could switch to `record`, though I'll either have to do manual changes to generated files (which I don't like), or modify UWP sync generator to add `record` as well (though I don't know if it should add it for all structs?)

Comment (jeromelaban) [reply src/Uno.UWP/ApplicationModel/PackageVersion.cs:5]: Ok, let's not modify the generator for this particular case. Adding the suggestions for linker friendliness is still a good change.

**Comment by Youssef1313** [comment]
@jeromelaban @spouliot Can you take another look please? Thanks!

File: src/Uno.Foundation/HResult.cs
Comment (spouliot) [inline src/Uno.Foundation/HResult.cs:15]: minor but, for the same reason (trimmability), it would be better to have `!=` call `!Equals` (that can't be trimmed) so the linker can eliminate `==` if unused.

File: src/Uno.UI.RemoteControl.Server.Processors/HotReload/MetadataUpdates/WatchHotReloadService.cs
Comment (spouliot) [inline src/Uno.UI.RemoteControl.Server.Processors/HotReload/MetadataUpdates/WatchHotReloadService.cs:43]: same

File: src/Uno.UI/Microsoft/UI/Xaml/Controls/Repeater/FlowLayoutAnchorInfo.cs
Comment (spouliot) [inline src/Uno.UI/Microsoft/UI/Xaml/Controls/Repeater/FlowLayoutAnchorInfo.cs:30]: same

File: src/Uno.UWP/Devices/Geolocation/BasicGeoposition.cs
Comment (spouliot) [inline src/Uno.UWP/Devices/Geolocation/BasicGeoposition.cs:24]: same

File: src/Uno.UWP/Gaming/Input/GamepadReading.cs
Comment (spouliot) [inline src/Uno.UWP/Gaming/Input/GamepadReading.cs:70]: same

File: src/Uno.UWP/UI/Input/ManipulationDelta.cs
Comment (spouliot) [inline src/Uno.UWP/UI/Input/ManipulationDelta.cs:79]: same

File: src/Uno.UWP/UI/Input/ManipulationVelocities.cs
Comment (spouliot) [inline src/Uno.UWP/UI/Input/ManipulationVelocities.cs:62]: same

**Review by spouliot — APPROVED** [review]

File: src/Uno.Foundation/HResult.cs
Comment (Youssef1313) [inline src/Uno.Foundation/HResult.cs:15]:
```suggestion
		public static bool operator !=(HResult left, HResult right) => !left.Equals(right);
```

File: src/Uno.UI.RemoteControl.Server.Processors/HotReload/MetadataUpdates/WatchHotReloadService.cs
Comment (Youssef1313) [inline src/Uno.UI.RemoteControl.Server.Processors/HotReload/MetadataUpdates/WatchHotReloadService.cs:43]:
```suggestion
			public static bool operator !=(Update left, Update right) => !left.Equals(right);
```

File: src/Uno.UI/Microsoft/UI/Xaml/Controls/Repeater/FlowLayoutAnchorInfo.cs
Comment (Youssef1313) [inline src/Uno.UI/Microsoft/UI/Xaml/Controls/Repeater/FlowLayoutAnchorInfo.cs:30]:
```suggestion
		public static bool operator !=(FlowLayoutAnchorInfo left, FlowLayoutAnchorInfo right) => !left.Equals(right);
```

File: src/Uno.UWP/Devices/Geolocation/BasicGeoposition.cs
Comment (Youssef1313) [inline src/Uno.UWP/Devices/Geolocation/BasicGeoposition.cs:24]:
```suggestion
		public static bool operator !=(BasicGeoposition left, BasicGeoposition right) => !left.Equals(right);
```

File: src/Uno.UWP/Gaming/Input/GamepadReading.cs
Comment (Youssef1313) [inline src/Uno.UWP/Gaming/Input/GamepadReading.cs:70]:
```suggestion
	public static bool operator !=(GamepadReading left, GamepadReading right) => !left.Equals(right);
```

File: src/Uno.UWP/UI/Input/ManipulationDelta.cs
Comment (Youssef1313) [inline src/Uno.UWP/UI/Input/ManipulationDelta.cs:79]:
```suggestion
		public static bool operator !=(ManipulationDelta left, ManipulationDelta right) => !left.Equals(right);
```

File: src/Uno.UWP/UI/Input/ManipulationVelocities.cs
Comment (Youssef1313) [inline src/Uno.UWP/UI/Input/ManipulationVelocities.cs:62]:
```suggestion
		public static bool operator !=(ManipulationVelocities left, ManipulationVelocities right) => !left.Equals(right);
```

File: src/Uno.UWP/ApplicationModel/PackageVersion.cs
Comment (Youssef1313) [inline src/Uno.UWP/ApplicationModel/PackageVersion.cs:42]:
```suggestion
		public static bool operator !=(PackageVersion left, PackageVersion right) => !left.Equals(right);
```

File: src/Directory.Build.props
Comment (agneszitte-nventive) [inline src/Directory.Build.props:205]:
```suggestion
        <!-- Enable performance CA rules from 'Microsoft.CodeAnalysis.NetAnalyzers' as build warnings by default. Specific rules are disabled or downgraded in the repo's editorconfig. -->
```

**Review by agneszitte-nventive — APPROVED** [review]

**Comment by Youssef1313** [comment]
@jeromelaban Could you restart the failing checks? Thanks!

**Comment by Youssef1313** [comment]
@jeromelaban There are remaining failures here :(

**Comment by jeromelaban** [comment]
Still in-between generators optimizations. Let's retry.

**Comment by Youssef1313** [comment]
@jeromelaban There are remaining failures here. Can you restart once again please? I hope this is ready to merge after that. Thanks!

**Comment by Youssef1313** [comment]
@jeromelaban This is green now! 🎉

File: src/Uno.UI.RemoteControl.Server.Processors/HotReload/MetadataUpdates/WatchHotReloadService.cs
Comment (jeromelaban) [inline src/Uno.UI.RemoteControl.Server.Processors/HotReload/MetadataUpdates/WatchHotReloadService.cs:23]: Overall, I find that the maintainability of this kind of changes pretty low. This particular change would benefit from being a record.

For the others, where records are a breaking change, would you know of an existing generator that would do that for structs (we had [one similar in Uno.CodeGen](https://github.com/unoplatform/Uno.CodeGen/blob/master/src/Uno.CodeGen/EqualityGenerator.cs))? Otherwise, I don't necessarily like regions but having one that explicitly says that the equality related members should be updated on any change would be interesting.

Comment (Youssef1313) [reply src/Uno.UI.RemoteControl.Server.Processors/HotReload/MetadataUpdates/WatchHotReloadService.cs:23]: I'll update `Update` struct to be a `record struct`.

For the others, I don't know of an existing generator, I'll have a region and a comment to note they should be updated for new fields.

**Review by MartinZikmund — APPROVED** [review]

**Pull request merged**
